docs/storage-config.md

```diff
@@ -0,0 +1,267 @@
+k8s which are not guaranteed / supported interfaces, so not desirable to do.
+
+At this time there is no way to make host block devices directly accessible to
+PODs. If this is enabled in the future, it will share the same problem as local
+file storage, in that the libvirtd POD cannot access block devices associated
+with the VM POD, so any block devices used would have to be devices that are
+visible to the libvirtd POD when it is initially launched.
+
+### Direct network block devices
+
+Given the complications with local file/block storage, the simplest way to
+configure a virtual disk is using a direct network block device. In the VM spec
+this can be configured two ways.
+
+#### Direct storage reference
+
+The storage server can be referenced directly, for example, as an iSCSI server
+address:
+
+```yaml
+    disks:
+      - source:
+          iscsi:
+            host:
+              name: iscsi-demo-target
+              port: "3260"
+            portal: iqn.2017-01.io.kubevirt:sn.42
+	    lun: 2
+```
+
+This would also work for RBD, NBD, GlusterFS, and HTTP(s) protocols.
+
+#### Indirect storage reference
+
+The k8s PersistentVolumeClaim (PVC) resource is an abstract concept that allows
+a tenant user to claim a PersistentVolume (PV). There are many different
+backends for PVs, some are filesystem based and some are network block device
+based. Using QEMU's direct network backend drivers, it is possible to directly
+access the storage for any PV that is backend by a network block device (in
+particular the iSCSI and RBD k8s backends).
+
+```yaml
+    disks:
+      - source:
+          pvc:
+	    claimname: some-pvc-name
+```
+
+When launching QEMU the PVC resource is accessed to get the corresponding
+PV resource. The configuration for the ISCSI / RBD backend is then extracted
+to configure the QEMU storage backend
+
+
+### Dynamic storage reference
+
+Rather than having the storage volume pre-created upfront, it may be useful to
+have a way to dynamically create storage backends. To achieve this, it is
+possible to define a storage source which defines a container. The virt-handler
+would create an instance of this container, which would be required to expose a
+HTTP service with a well-known URI eg https://$POD-IP/volume
+
+Making a PUT request to this URI would trigger creation of the volume, and a
+a DELETE request would naturally delete it. The response to the PUT request
+would be a YAML document following the 'iscsi' or 'pvc' schema for the VM
+spec disk source. eg
+
+```yaml
+    disks:
+      - source:
+          container:
+	    image: some-image-name
+	    ... any other k8s container attribute...
+```
+
+Upon making the PUT request it would respond with
+
+```yaml
+    pvc:
+      claimname: some-pvc-name
+```
+
+or
+
+```yaml
+    iscsi:
+      host:
+        name: iscsi-demo-target
+        port: "3260"
+      portal: iqn.2017-01.io.kubevirt:sn.42
+    lun: 2
+```
+
+## Access modes
```

Comment:
QEMU can enforce access mode on storage that is file based, but not currently on network based storage. It would be technically possible to enforce it on iSCSI based storage using SCSI reservations, but that's not implemented right now.  Libvirt also has ability to perform locking on disks volumes, and has a number of options in this area. Again currently, it doesn't do it for network based storage, but it is technically possible to extend it. So in general access mode concept makes sense.